### PR TITLE
Above ground zoi

### PR DESCRIPTION
**Pull request by zaphod-bx**

### CODE REVIEW

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    ## This function calculates the tree height at a (mesh-)point depending
+    #  on the distance from the tree position.\n
+    #  @param stem_height - stem height\n
+    #  @param crown_radius - crown radius\n
+    #  @param distance - distance from the stem position
+    def calculateHeightFromDistance(self,stem_height,crown_radius,distance):
+        height = crown_radius-distance
+        height = np.maximum(height, 0,height)
+        height[height>0] = stem_height + (4*crown_radius**2 - distance[height>0]**2)**0.5
+        return height
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: As before in pr #14  : some suggestions for performance improvements.
the timing was just added for developping purposes and can be removed









`   
 ## This function returns the AbovegroundResources calculated in the
    #  subsequent timestep.\n
    #  @return: np.array with $N_tree$ scalars
    def calculateAbovegroundResources(self):
        #for i in range(len(self.xe)):
        t_start = time.time()
        distance = (((self.my_grid[0][:,:,np.newaxis] - np.array(self.xe)[np.newaxis,np.newaxis,:])**2
                     + (self.my_grid[1][:,:,np.newaxis]-np.array(self.ye)[np.newaxis,np.newaxis,:])**2)**0.5)
        t_dist = time.time()
        my_height, canopy_bools = self.calculateHeightFromDistance(np.array(self.h_stem),np.array(self.r_crown),distance)
        t_height = time.time()
        #Define empty array of dimensions res_x, res_y, n_trees
        crown_areas = np.zeros_like(my_height)
        #Add a one, where tree is larger than 0
        crown_areas[np.where(canopy_bools)] +=1
        #Count all nodes, which are occupied by trees
        #returns array of shape (ntrees)
        crown_areas = crown_areas.sum(axis=(0,1))

        # look for largest tree
        self.canopy_height = np.max(my_height, axis=-1)
        self.canopy_height[np.where(np.less(self.canopy_height,0))] = 0
        #print(np.shape(my_height.sum(axis=-1)))
        #print(np.shape(my_height.sum(axis=0)))
        #crown_area.append(sum(sum(my_height > 0)))

        #define array to count wins
        wins = np.zeros_like(my_height)
        #indicate, where tree is highest
        wins[np.where(np.equal(my_height,self.canopy_height[:,:,np.newaxis]))] +=1

        #Account for shared wins
        cumwins = wins.sum(axis=-1)
        cumwins[np.where(cumwins==0)] = -1e-4
        wins = wins/ cumwins[:,:,np.newaxis]
        wins[np.where(wins<0)] = 0
        #Count number of wins
        wins=wins.sum(axis=(0,1))
        t_end = time.time()
        self.aboveground_resources = wins/crown_areas
        print("t_dist:", (t_dist - t_start)," in %", (t_dist - t_start)/(t_end-t_start))
        print("t_h_alloc:", (t_height - t_dist)," in %", (t_height - t_dist)/(t_end-t_start))
        print("t_dist:", (t_end - t_height)," in %", (t_end - t_height)/(t_end-t_start))


    ## This function calculates the tree height at a (mesh-)point depending
    #  on the distance from the tree position.\n
    #  @param hst - stem height\n
    #  @param rcr - crown radius\n
    #  @param distance - distance from the stem position
    def calculateHeightFromDistance(self,hst,rcr,distance):
        bools = rcr[np.newaxis, np.newaxis,:]>distance
        idx = np.where(bools)
        height = np.full_like(distance, fill_value = -99999)
        height[idx] = hst[idx[2]] + (4*rcr[idx[2]]**2 - distance[idx]**2)**0.5
        return height, bools
`

**Comment by jbathmann** [comment]
Nice.
Please run: yapf -r -i ./ 
for formatting and maybe incorporate suggestion

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 1 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: soll ich das jetzt ersetzen? Wenn das so funktioniert, bin ich dafür.

**Comment by zaphod-bx** [comment]
yes, geyapft. Das mergen vorhin war aus Versehen, ich hatte noch gar nix wieder gepusht.

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 1 above)
Comment (jbathmann) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Wenn du magst gerne. Vielleicht kannst du ja die Zeitausgaben wieder raus nehmen, rsc und hsc wieder ersetzen, die jetzt überflüssig gewordenen initialisierungen rausnehmen und das ganze nochmal yapfen? :D
Ich hoffe du kannst es nachvollziehen?

Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: halbwegs.

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    ## This functions prepares arrays for the competition
+    #  concept. In the SimpleAssymmetricZOI concept, trees geometric measures
+    #  are saved in simple lists and the timestepping is updated. Two numpy
+    #  arrays similar to the mesh array for canopy height and tree ID of highest tree
+    #  at corresponding mesh point are initialised. \n
+    #  @param t_ini - initial time for next timestep \n
+    #  @param t_end - end time for next timestep
+    def prepareNextTimeStep(self, t_ini, t_end):
+        self.crown_area = []
+        self.tree_win = []
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Those can be removed

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        self.xe = []
+        self.ye = []
+        self.h_stem = []
+        self.r_crown = []
+        self.t_ini = t_ini
+        self.t_end = t_end
+        self.canopy_height = np.zeros_like(self.my_grid[0])
+        self.winner = np.full_like(self.my_grid[0], fill_value=np.nan)
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Überflüssig

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        #Define empty array of dimensions res_x, res_y, n_trees
+        crown_areas = np.zeros_like(my_height)
+        #Add a one, where tree is larger than 0
+        crown_areas[np.where(canopy_bools)] += 1
+        #Count all nodes, which are occupied by trees
+        #returns array of shape (ntrees)
+        crown_areas = crown_areas.sum(axis=(0, 1))
+
+        # look for largest tree
+        self.canopy_height = np.max(my_height, axis=-1)
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: no "self" assignment needed here

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                string += tag + " "
+            raise KeyError(
+                "Tag(s) " + string +
+                "are not given for above-ground grid initialisation in project file."
+            )
+        l_x = x_2 - x_1
+        l_y = y_2 - y_1
+        x_step = l_x / x_resolution
+        y_step = l_y / y_resolution
+        self.min_r_crown = np.mean([x_step, y_step]) * 2**0.5
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Isn't half the size sufficient to at least get one node?
I would rather use np.ax, cause it's more robust

`
```suggestion
        self.min_r_crown = np.max([x_step, y_step]) * 1/2**0.5
```

`

**Comment by jbathmann** [comment]
Almost there, this is going to be awesome. ;-)

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 5 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: I thought we want at least 3?
and shouldn't it be np.min, then?

Comment (jbathmann) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Why is 3 better than one?
np.max in order to capture the largest spacing. Imagine x_step = 0.1 and y_step = 10. In that case we a min crown radius of 0.1/sqrt(2) does not imply a minimum of 1 (or 3) gridpoints.

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                string += tag + " "
+            raise KeyError(
+                "Tag(s) " + string +
+                "are not given for above-ground grid initialisation in project file."
+            )
+        l_x = x_2 - x_1
+        l_y = y_2 - y_1
+        x_step = l_x / x_resolution
+        y_step = l_y / y_resolution
+        self.min_r_crown = np.min([x_step, y_step]) * 1 / 2**0.5
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Why np.min? Überzeug mich :dancer:

**Pull request merged**